cmd/geth/passwords.go

```diff
@@ -0,0 +1,154 @@
+	for flag, val := range vaultFlags {
+		if val == "" {
+			missingFlags = append(missingFlags, flag.Name)
+		}
+	}
+	switch len(missingFlags) {
+	case 0:
+		// Ensure there were no other password args before returning true.  Much safety, very check.
+		passwordFlags := map[cli.StringFlag]string{
+			utils.VoteAccountPasswordFlag:           strings.TrimSpace(ctx.GlobalString(utils.VoteAccountPasswordFlag.Name)),
+			utils.VoteBlockMakerAccountPasswordFlag: strings.TrimSpace(ctx.GlobalString(utils.VoteBlockMakerAccountPasswordFlag.Name)),
+			utils.PasswordFileFlag:                  strings.TrimSpace(ctx.GlobalString(utils.PasswordFileFlag.Name)),
+		}
+		setPassFlags := make([]string, 0)
+		for flag, val := range passwordFlags {
+			if val != "" {
+				setPassFlags = append(setPassFlags, flag.Name)
+			}
+		}
+		if len(setPassFlags) > 0 {
+			utils.Fatalf("Collision: both Vault and other password flags were specified. If you are using Vault, these should not be present: %v", setPassFlags)
+		}
+		return true
+	case 1:
+		// Bad case, have one but missing another, throw an error
+		// and let 'em know what's missing
+		utils.Fatalf("Some Vault flags specified, but not enough to retrieve the password; please include: %v", missingFlags)
+		return true
+	case 2:
+		// Vanilla case, as two of these have default values
+		return false
+	default:
+		utils.Fatalf("Unexpected number of Vault args missing, two of four should always be specified via defaults.")
+		return false
+	}
+}
+
+// Expects to be running in EC2
+func getIAMRole() (string, error) {
+	svc := ec2metadata.New(session.New())
+	iam, err := svc.IAMInfo()
+	if err != nil {
+		return "", err
+	}
+	// Our instance profile conveniently has the same name as the role
+	profile := iam.InstanceProfileArn
+	splitArn := strings.Split(profile, "/")
+	if len(splitArn) < 2 {
+		return "", fmt.Errorf("No / character found in instance profile ARN")
+	}
+	role := splitArn[1]
+	return role, nil
+}
+
+func loginAws(v *vaultAPI.Client) (string, error) {
+	// Login data args are left empty so that Vault's AWSAuth implementation
+	// knows to let AWS's EnvProvider handle it.  The EnvProvider searches the
+	// environment for these values: https://github.com/aws/aws-sdk-go/blob/master/aws/credentials/env_provider.go
+	loginData, err := awsauth.GenerateLoginData( /*accessKey=*/ "" /*secretKey=*/, "" /*sessionToken=*/, "" /*headerValue=*/, "")
+	if err != nil {
+		return "", err
+	}
+	if loginData == nil {
+		return "", fmt.Errorf("Got nil response from GenerateLoginData")
+	}
+
+	role, err := getIAMRole()
+	if err != nil {
+		return "", err
+	}
+	loginData["role"] = role
+
+	path := "auth/aws/login"
+
+	secret, err := v.Logical().Write(path, loginData)
+	if err != nil {
+		return "", err
+	}
+	if secret == nil {
+		return "", fmt.Errorf("Empty secret response from credential provider.")
+	}
+	if secret.Auth == nil {
+		return "", fmt.Errorf("Secret contains no auth data.")
+	}
+	if secret.Auth.ClientToken == "" {
```

Comment:
Do we also need to check for nil?

Reply:
We actually can't check for nil, as that doesn't satisfy the type requirement of `ClientToken`.  I added this `""` check because a nil check wouldn't compile.